rfcs/side-effect-lifecycle-scheduling.md

```diff
@@ -0,0 +1,124 @@
+        .add_system(system_that_applies_state_queue
+            .consumes("state change")
+            .produces("state")
+        )
+        .add_system(run_criteria
+            .reads("state")
+            .produces("on update state=A")
+        )
+        .add_system(system_A.reads("on update state=A"))
+        .add_system(system_B.reads("on update state=A"))
+}
+```
+
+## Implementation
+
+If we tried to create the relationships between producers, consumers, readers, and writers with system sets and labeling, it would look something like this.
+```rs
+// expressing the subgraph with system sets and labels
+
+fn main() {
+    let SideEffectLabel = Label::new("some side effect");
+    App::new()
+        .add_system_set(SystemSet::new()
+            .label(SideEffectLabel)
+            .add_system_set(SystemSet::new()
+                .label(produces(SideEffectLabel))
+                .before(reads(SideEffectLabel))
+                .before(writes(SideEffectLabel))
+                .before(consumes(SideEffectLabel))
+                .add_system(side_effect_producer)
+            )
+        )
+        // TODO: add consumers, readers, and writer system sets
+}
+```
+
+In practice we can't create these system sets like this as it would be very unergonomic for users. But systems sets are just sugar anyways and don't exist at runtime. We can just apply the correct labels to individual systems.
+
+```rs
+// .produces
+.add_system(side_effect_producer.produces(SideEffectLabel))
+// is equivalent to
+.add_system(side_effect_producer
+    .label(SideEffectLabel)
+    .label(produces(SideEffectLabel))
+    .before(reads(SideEffectLabel))
+    .before(writes(SideEffectLabel))
+    .before(comsumes(SideEffectLabel))
+)
+```
+
+> TODO: add desugaring of readers, writers, and consumers.
+
+## Drawbacks
+
+I expect that these relationships will mostly be added automatically when using run criteria, app state, and the like with a higher level api. Users would only be expected to interact with this API when things can't be done automatically or the default behavior doesn't match their needs.
+
+## Open questions
+
+* With looping criteria it might be possible to create a valid dependency cycle that cannot be resolved in the current API. If that is the case then we might need an `.after_can_break` type of edge that could be ignored in dependency cycle checking.
+
+## Alternatives Considered
+
+* The example above for expressing the relationships between side effects using just labeling and system sets has one major problem. There is no current api for adding systems to system sets after the system set has been created. If such an api did exist we could potentially just add systems to the correct system set. This didn't seem better than the proposed API and has some implementation complexity since system sets are mostly just api sugar for labeling multiple systems and does not exist at runtime.
+
+## Future Work
```

Comment:
Indexing and relations are two other highly visible use cases that are enabled by solving the data consistency problem.